Improve aten::backward handling

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36750 Improve aten::backward handling**

- It seems the JIT schema for aten::backward and the schema in native_functions.yaml diverged on whether the retain_graph/keep_graph parameter takes a `bool` or a `bool?`. Make them identical again.
- Also remove the mutability annotation for the self parameter. This does not make sense together with AliasAnalysisKind::CONSERVATIVE and it triggered an assertion
- After fixing the mutability annotation, we can fix that assertion so that it doesn't exclude aten::backward from its check anymore
- Last but not least, remove the unboxed_only marker from aten::backward. This requires us to add a special case in register_c10_ops.cpp for it, because JIT still has its own implementation.


After this PR, we still have separate implementations for aten::backward for JIT (in register_prim_ops_fulljit.cpp) and for eager mode (in native_functions.yaml), but they share the same signature again.

Differential Revision: [D21004102](https://our.internmc.facebook.com/intern/diff/D21004102/)

smessmer committed:
Improve aten::backward handling

- It seems the JIT schema for aten::backward and the schema in native_functions.yaml diverged on whether the retain_graph/keep_graph parameter takes a `bool` or a `bool?`. Make them identical again.
- Also remove the mutability annotation for the self parameter. This does not make sense together with AliasAnalysisKind::CONSERVATIVE and it triggered an assertion
- After fixing the mutability annotation, we can fix that assertion so that it doesn't exclude aten::backward from its check anymore
- Last but not least, remove the unboxed_only marker from aten::backward. This requires us to add a special case in register_c10_ops.cpp for it, because JIT still has its own implementation.

Differential Revision: [D21004102](https://our.internmc.facebook.com/intern/diff/D21004102/)

[ghstack-poisoned]

smessmer committed:
Update on "Improve aten::backward handling"


- It seems the JIT schema for aten::backward and the schema in native_functions.yaml diverged on whether the retain_graph/keep_graph parameter takes a `bool` or a `bool?`. Make them identical again.
- Also remove the mutability annotation for the self parameter. This does not make sense together with AliasAnalysisKind::CONSERVATIVE and it triggered an assertion
- After fixing the mutability annotation, we can fix that assertion so that it doesn't exclude aten::backward from its check anymore
- Last but not least, remove the unboxed_only marker from aten::backward. This requires us to add a special case in register_c10_ops.cpp for it, because JIT still has its own implementation.

Differential Revision: [D21004102](https://our.internmc.facebook.com/intern/diff/D21004102/)

[ghstack-poisoned]

bhosmer commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -117,7 +117,7 @@ RegisterOperators reg(
      // create_graph=True so we use aliasAnalysisConservative for these two OPs
      Operator(
-         "aten::backward(Tensor[](a!) tensors, Tensor?[]? grad_tensors=None, bool? retain_graph=None, bool create_graph=False) -> ()",
+         "aten::backward.TensorList(Tensor[] tensors, Tensor?[]? grad_tensors=None, bool? retain_graph=None, bool create_graph=False) -> ()",
```
This was name duplicated before? It's serialization of calls to the unlucky variant that would break, right?

smessmer replied:
yes but that serialization only happens for the lite interpreter where we don't have bc concerns for models doing autograd yet.

bhosmer reviewed (approved):
LG mod tests :) 

Maybe add a sentence summarizing the post-PR end state to the description? the changes make sense but a recap of the "improved" situation might be good

smessmer committed:
Update on "Improve aten::backward handling"


- It seems the JIT schema for aten::backward and the schema in native_functions.yaml diverged on whether the retain_graph/keep_graph parameter takes a `bool` or a `bool?`. Make them identical again.
- Also remove the mutability annotation for the self parameter. This does not make sense together with AliasAnalysisKind::CONSERVATIVE and it triggered an assertion
- After fixing the mutability annotation, we can fix that assertion so that it doesn't exclude aten::backward from its check anymore
- Last but not least, remove the unboxed_only marker from aten::backward. This requires us to add a special case in register_c10_ops.cpp for it, because JIT still has its own implementation.


After this PR, we still have separate implementations for aten::backward for JIT (in register_prim_ops_fulljit.cpp) and for eager mode (in native_functions.yaml), but they share the same signature again.

Differential Revision: [D21004102](https://our.internmc.facebook.com/intern/diff/D21004102/)

[ghstack-poisoned]

smessmer committed:
Update on "Improve aten::backward handling"


- It seems the JIT schema for aten::backward and the schema in native_functions.yaml diverged on whether the retain_graph/keep_graph parameter takes a `bool` or a `bool?`. Make them identical again.
- Also remove the mutability annotation for the self parameter. This does not make sense together with AliasAnalysisKind::CONSERVATIVE and it triggered an assertion
- After fixing the mutability annotation, we can fix that assertion so that it doesn't exclude aten::backward from its check anymore
- Last but not least, remove the unboxed_only marker from aten::backward. This requires us to add a special case in register_c10_ops.cpp for it, because JIT still has its own implementation.


After this PR, we still have separate implementations for aten::backward for JIT (in register_prim_ops_fulljit.cpp) and for eager mode (in native_functions.yaml), but they share the same signature again.

Differential Revision: [D21004102](https://our.internmc.facebook.com/intern/diff/D21004102/)

[ghstack-poisoned]
